File: iOS/JSBridge/JSBridge/JSBridge/JSBridge.m
//
//  JSBridge.m
//  JSBridge
//
//  Created by Siva RamaKrishna Ravuri
//  Copyright (c) 2014 www.siva4u.com. All rights reserved.
//
// The MIT License (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
//

#import <objc/runtime.h>
#import "JSBridge.h"
#import "JSBridgeBase.h"

@interface JSBridge()
@property(nonatomic,assign) UIWebView               *jsWebView;
@property(nonatomic,assign) id<UIWebViewDelegate>   jsWebViewDelegate;
@property(nonatomic,assign) NSBundle                *resourceBundle;
@property(nonatomic,assign) JSBHandler              bridgeHandler;
@property(nonatomic,assign) long                    uniqueId;
@property(nonatomic,assign) NSUInteger              numberOfUrlRequests;

@property(nonatomic,retain) NSMutableArray          *startupMessageQueue;
@property(nonatomic,retain) NSMutableDictionary     *responseCallbacks;
@property(nonatomic,retain) NSMutableDictionary     *messageHandlers;
@property(nonatomic,retain) NSMutableDictionary     *nativeModules;
@end

@implementation JSBridge

@synthesize jsWebView;
@synthesize jsWebViewDelegate;
@synthesize resourceBundle;
@synthesize bridgeHandler;
@synthesize uniqueId;
@synthesize numberOfUrlRequests;

@synthesize startupMessageQueue;
@synthesize responseCallbacks;
@synthesize messageHandlers;
@synthesize nativeModules;

#pragma mark - Alloc-Dealloc

-(void)initialize {
    jsWebViewDelegate   = nil;
    resourceBundle      = nil;
    bridgeHandler       = nil;
    uniqueId            = 0;
    numberOfUrlRequests = 0;
    
    if(jsWebView) jsWebView.delegate = nil;
    RELEASE_MEM(jsWebView);
    RELEASE_MEM(startupMessageQueue);
    RELEASE_MEM(responseCallbacks);
    RELEASE_MEM(messageHandlers);
    RELEASE_MEM(nativeModules);
}

-(void)dealloc {
    [self initialize];
#if __has_feature(objc_arc)
#else
    [super dealloc];
#endif
}

#pragma mark - PRIVATE APIs

-(void)dispatchMessage:(NSDictionary *)message {
    NSString *messageJSON = [JSBridge stringifyJSON:message];
    JSBLog(@"JSB Action: SEND: %@",messageJSON);
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\\" withString:@"\\\\"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\"" withString:@"\\\""];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\'" withString:@"\\\'"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\n" withString:@"\\n"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\r" withString:@"\\r"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\f" withString:@"\\f"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\u2028" withString:@"\\u2028"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\u2029" withString:@"\\u2029"];
    
    NSString* javascriptCommand = [NSString stringWithFormat:@"%@.%@('%@');",JS_BRIDGE,JS_BRIDGE_SEND_NATIVE_QUEUE,messageJSON];
    if ([[NSThread currentThread] isMainThread]) {
        [jsWebView stringByEvaluatingJavaScriptFromString:javascriptCommand];
    } else {
        __strong UIWebView* strongWebView = jsWebView;
        dispatch_sync(dispatch_get_main_queue(), ^{
            [strongWebView stringByEvaluatingJavaScriptFromString:javascriptCommand];
        });
    }
}

-(void)queueMessage:(NSDictionary *)message {
    if (startupMessageQueue) {
        [startupMessageQueue addObject:message];
    } else {
        [self dispatchMessage:message];
    }
}

-(NSObject *)getNativeModuleFromName:(NSString *)name webView:(UIWebView *)webView {
    NSObject *nativeModule	= [nativeModules objectForKey:name];
    if(nativeModule == nil) {
        Class objClass = NSClassFromString(name);
        if(objClass) {
            @try {
                nativeModule = [[objClass alloc] initWithJSBridge:self webView:webView];
                [nativeModules setObject:nativeModule forKey:name];
                }
            @catch (NSException *exception) {
                JSBLog(@"getNativeModuleFromName: EXCEPTION: %@",name);
                nativeModule = nil;
            }
            @finally {
            }
        } else {
            JSBLog(@"Unsupported Module: %@",name);
        }
    }
    return nativeModule;
}

-(void)handleReturnValue:(NSInvocation *)invoker sig:(NSMethodSignature *)sig webView:(UIWebView *)webView apiName:(NSString *)apiName status:(BOOL)status {
    NSString *retValue = nil;
    if((invoker != nil) && (sig != nil)) {
        if([sig methodReturnLength] > 0) {
            [invoker getReturnValue:&retValue];
            if(retValue) {
                JSBLog(@"handleReturnValue:%@",retValue);
                retValue = (NSString *)CFBridgingRelease(CFURLCreateStringByAddingPercentEscapes(NULL,(CFStringRef) retValue, NULL, (CFStringRef)@"!*'();:@&=+$,/?%#[]", kCFStringEncodingUTF8));
            }
        }
    }
    
    NSString *dataStr = @"";
    if(status == false) retValue = [NSString stringWithFormat:@"UN-SUPPORTED API: %@",apiName];
    if(retValue) dataStr = [NSString stringWithFormat:@",'data':'%@'",retValue];
    [webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"JSBridge.nativeReturnValue = \"{'status':'%@'%@}\"",((status)?(@"true"):(@"false")),dataStr]];
    
    RELEASE_MEM(retValue);
}

-(void)processEventHandler:(UIWebView *)webView message:(NSDictionary *)message responseCallback:(JSBResponseCallback)responseCallback {
    NSString *eventName = message[@"eventName"];
    if(eventName) {
        JSBHandler handler = messageHandlers[eventName];
        if(!handler) {
            @try {
                // eventName is not registered and so create an instance of the API
                NSArray *api        = [eventName componentsSeparatedByString:@"."];
                NSObject *jsModule  = [self getNativeModuleFromName:(NSString*)[api objectAtIndex:0] webView:webView];
                if(jsModule) {
                    SEL selector                = NSSelectorFromString([NSString stringWithFormat:@"JSBEvent_%@:responseCallback:",(NSString*)[api objectAtIndex:1]]);
                    NSMethodSignature *sig      = [[jsModule class] instanceMethodSignatureForSelector:selector];
                    if(sig) {
                        NSInvocation *invoker   = [NSInvocation invocationWithMethodSignature:sig];
                        invoker.selector        = selector;
                        invoker.target          = jsModule;
                        [self registerEvent:eventName handler:^(id data, JSBResponseCallback responseCallback) {
                            NSDictionary *configData = message[@"data"];
                            if(configData) [invoker setArgument:&configData atIndex:2];
                            if(responseCallback) [invoker setArgument:&responseCallback atIndex:3];
                            [invoker invoke];
                        }];
                        
                        handler = messageHandlers[eventName];
                    } else {
                        JSBLog(@"processEventHandler: EXCEPTION: Unsupported Event: %@",eventName);
                    }
                } else {
                    JSBLog(@"processEventHandler: EXCEPTION: No Plugin: %@",eventName);
                }
            }
            @catch (NSException *exception) {
                JSBLog(@"processEventHandler: EXCEPTION: %@",eventName);
                handler = nil;
            }
            @finally {
            }
        }
        if(handler == nil) {
            handler = ^(id data, JSBResponseCallback responseCallback) {
                if(responseCallback) {
                    responseCallback(@{@"status":@false,@"data":[NSString stringWithFormat:@"UN-SUPPORTED EVENT: %@",eventName]});
                }
            };
        }
        handler(message[@"data"], responseCallback);
    } else {
        if(bridgeHandler) {
            bridgeHandler(message[@"data"], responseCallback);
        } else {
            JSBLog(@"EXCEPTION: No handler for message from JS: %@",message);
        }
    }
}

-(void)processJSEventQueue:(UIWebView *)webView {
    NSString *messageQueueString = [webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"%@.%@();",JS_BRIDGE,JS_BRIDGE_GET_JS_EVENT_QUEUE]];
    
    id messages = [JSBridge parseJSONArray:messageQueueString];
    if(![messages isKindOfClass:[NSArray class]]) {
        JSBLog(@"flushMessageQueue: WARNING: Invalid %@ received: %@", [messages class], messages);
        return;
    }
    
    for (NSDictionary *message in messages) {
        if (![message isKindOfClass:[NSDictionary  class]]) {
            JSBLog(@"flushMessageQueue: WARNING: Invalid %@ received: %@", [message class], message);
            continue;
        }
        JSBLog(@"flushMessageQueue: RCVD: %@",message);
        NSString* responseId = message[@"responseId"];
        if (responseId) {
            JSBResponseCallback responseCallback = responseCallbacks[responseId];
            responseCallback(message[@"responseData"]);
            [responseCallbacks removeObjectForKey:responseId];
        } else {
            JSBResponseCallback responseCallback = NULL;
            NSString* callbackId = message[@"callbackId"];
            if (callbackId) {
                responseCallback = ^(id responseData) {
                    if (responseData == nil) {
                        responseData = [NSNull null];
                    }
                    
                    NSDictionary *msg = @{ @"responseId":callbackId, @"responseData":responseData };
                    [self queueMessage:msg];
                };
            } else {
                responseCallback = ^(id ignoreResponseData) {
                    // Do nothing
                };
            }
            
            [self processEventHandler:webView message:message responseCallback:responseCallback];
        }
    }
}

-(void)processJSAPIRequest:(UIWebView *)webView {
    
    NSDictionary *cData = [JSBridge parseJSON:[webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"%@.%@();",JS_BRIDGE,JS_BRIDGE_GET_API_DATA]]];
    NSString *apiName   = [cData objectForKey:@"api"];
    @try {
        // execute the interfacing method
        NSArray  *api       = [apiName componentsSeparatedByString:@"."];
        NSObject *jsModule  = [self getNativeModuleFromName:(NSString*)[api objectAtIndex:0] webView:webView];
        if(jsModule) {
            SEL selector            = NSSelectorFromString([NSString stringWithFormat:@"JSBAPI_%@",(NSString*)[api objectAtIndex:1]]);
            NSMethodSignature *sig  = [[jsModule class] instanceMethodSignatureForSelector:selector];
            NSInvocation *invoker   = [NSInvocation invocationWithMethodSignature:sig];
            invoker.selector        = selector;
            invoker.target          = jsModule;
            
            NSString *apiDataStr    = [cData objectForKey:@"data"];
            JSBLog(@"JSB API: RCVD: %@(%@)",api,apiDataStr);
            if(apiDataStr) {
                NSDictionary *apiData = [JSBridge parseJSON:apiDataStr];
                [invoker setArgument:&apiData atIndex:2];
            }
            [invoker invoke];
            
            [self handleReturnValue:invoker sig:sig webView:webView apiName:apiName status:true];
            RELEASE_MEM(invoker);
        } else {
            [self handleReturnValue:nil sig:nil webView:webView apiName:apiName status:false];
        }
    }
    @catch (NSException *exception) {
        JSBLog(@"processJSAPIRequest: EXCEPTION: %@",exception);
        [self handleReturnValue:nil sig:nil webView:webView apiName:apiName status:false];
    }
    @finally {
    }
}

#pragma mark - WebView Delegates

-(void)webViewDidStartLoad:(UIWebView *)webView {
    if(webView != jsWebView) return;
    numberOfUrlRequests++;

    [jsWebView stringByEvaluatingJavaScriptFromString:@"window.isHybridMode = true"];

    __strong NSObject<UIWebViewDelegate>* strongDelegate = jsWebViewDelegate;
    if (strongDelegate && [strongDelegate respondsToSelector:@selector(webViewDidStartLoad:)]) {
        [strongDelegate webViewDidStartLoad:webView];
    }
}

-(void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
    if(webView != jsWebView) return;
    numberOfUrlRequests--;
    
    __strong NSObject<UIWebViewDelegate>* strongDelegate = jsWebViewDelegate;
    if (strongDelegate && [strongDelegate respondsToSelector:@selector(webView:didFailLoadWithError:)]) {
        [strongDelegate webView:webView didFailLoadWithError:error];
    }
}

-(void)webViewDidFinishLoad:(UIWebView *)webView {
    if(webView != jsWebView) return;
    numberOfUrlRequests--;
    
    if(numberOfUrlRequests == 0) {
        if(![[webView stringByEvaluatingJavaScriptFromString:[NSString stringWithFormat:@"typeof %@ == 'object'",JS_BRIDGE]] isEqualToString:@"true"]) {
            NSBundle *bundle = resourceBundle ? resourceBundle : [NSBundle mainBundle];
            NSString *filePath = [bundle pathForResource:JS_BRIDGE_FILE_NAME ofType:@"js"];
            NSString *js = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
            [webView stringByEvaluatingJavaScriptFromString:js];
        }
    }
    
    if (startupMessageQueue) {
        for (id queuedMessage in startupMessageQueue) {
            [self dispatchMessage:queuedMessage];
        }
        startupMessageQueue = nil;
    }
    
    __strong NSObject<UIWebViewDelegate>* strongDelegate = jsWebViewDelegate;
    if (strongDelegate && [strongDelegate respondsToSelector:@selector(webViewDidFinishLoad:)]) {
        [strongDelegate webViewDidFinishLoad:webView];
    }
}

-(BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {

    JSBLog(@"Received: %@",request);

    if(webView != jsWebView) return YES;
    
    NSURL *url = [request URL];
    if ([[url scheme] isEqualToString:JSBRIDGE_URL_SCHEME]) {
        if ([[url host] isEqualToString:JSBRIDGE_URL_MESSAGE]) {
            NSString *relativePath = [url relativePath];
            if([relativePath isEqualToString:JSBRIDGE_URL_EVENT_REL_PATH]) {
                [self processJSEventQueue:webView];
            } else if([relativePath isEqualToString:JSBRIDGE_URL_API_REL_PATH]) {
                [self processJSAPIRequest:webView];
            }
        } else {
            JSBLog(@"shouldStartLoadWithRequest: WARNING: Received unknown command %@",url);
        }
        return NO;
    } else {
        __strong NSObject<UIWebViewDelegate>* strongDelegate = jsWebViewDelegate;
        if (strongDelegate && [strongDelegate respondsToSelector:@selector(webView:shouldStartLoadWithRequest:navigationType:)]) {
            return [strongDelegate webView:webView shouldStartLoadWithRequest:request navigationType:navigationType];
        }
    }
    return YES;
}

#pragma mark - PRIVATE STATIC APIs

+(NSArray*)parseJSONArray:(NSString *)messageJSON {
    return [NSJSONSerialization JSONObjectWithData:[messageJSON dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingAllowFragments error:nil];
}

+(NSDictionary *)getReturnObjectWithStatus:(BOOL)status apiName:(NSString *)apiName data:(id)data {
    NSMutableDictionary *retValue = [[NSMutableDictionary alloc] initWithObjectsAndKeys:((status)?(@"true"):(@"false")),@"status",nil];
    if(apiName) [retValue setObject:apiName forKey:@"apiName"];
    if(data) [retValue setObject:data forKey:@"data"];
    return retValue;
}
+(NSDictionary *)getReturnObjectWithStatus:(BOOL)status data:(id)data {
    return [JSBridge getReturnObjectWithStatus:status apiName:nil data:data];
}

+(void)callCallbackWithStatus:(UIWebView *)wv inJSON:(NSDictionary *)inJSON outJSON:(id)outJSON status:(BOOL)status {
    if(inJSON) {
        NSString *callbackID = [inJSON objectForKey:@"callbackID"];
        if(callbackID) {
            
            NSString *removeAfterExecute = [inJSON objectForKey:@"removeAfterExecute"];
            if(!removeAfterExecute) removeAfterExecute = @"true";
            
            NSDictionary *retObj     = [JSBridge getReturnObjectWithStatus:status data:outJSON];
            NSString *retVal         = [JSBridge stringifyJSON:retObj];
            NSString *jsAPIToExecute = [NSString stringWithFormat:@"JSBridge._invokeJSCallback(\"%@\", %@, %@);",callbackID,removeAfterExecute,retVal];
            [wv stringByEvaluatingJavaScriptFromString:jsAPIToExecute];
        }
    }
}

+(void)callEventCallback:(JSBResponseCallback)responseCallback data:(NSDictionary *)data status:(BOOL)status {
    if(responseCallback != nil) {
        responseCallback([JSBridge getReturnObjectWithStatus:status data:data]);
    }
}

#pragma mark - PUBLIC STATIC APIs

+(NSString *)stringifyJSON:(id)message {
    if([NSJSONSerialization isValidJSONObject:message]) {
        return [[NSString alloc] initWithData:[NSJSONSerialization dataWithJSONObject:message options:0 error:nil] encoding:NSUTF8StringEncoding];
    }
    return @"";
}

+(NSDictionary*)parseJSON:(NSString *)messageJSON {
    return [NSJSONSerialization JSONObjectWithData:[messageJSON dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingAllowFragments error:nil];
}

+(NSString *)getString:(NSString *)str {
    return ((str)?(str):(@""));
}

+(NSDictionary *)putKeyValue:(NSMutableDictionary *)src key:(NSString *)key value:(id)value {
    if(src == nil) src = [[NSMutableDictionary alloc]init];
    if((key != nil) && (value != nil)) [src setObject:value forKey:key];
    return src;
}

+(NSDictionary *)getReturnObject:(id)data {
    return [JSBridge getReturnObjectWithStatus:true data:data];
}

+(void)callCallbackForWebView:(UIWebView *)wv inJSON:(NSDictionary *)inJSON outJSON:(id)outJSON {
    [JSBridge callCallbackWithStatus:wv inJSON:inJSON outJSON:outJSON status:true];
}

+(void)callEventCallback:(JSBResponseCallback)responseCallback data:(id)data {
    [JSBridge callEventCallback:responseCallback data:data status:true];
}

#pragma mark - PUBLIC APIs

-(id)initWithWebView:(UIWebView*)webView webViewDelegate:(NSObject<UIWebViewDelegate>*)webViewDelegate bundle:(NSBundle*)bundle handler:(JSBHandler)handler {
    self = [super init];
    if(self) {
        [self initialize];
        jsWebView           = webView;
        jsWebView.delegate  = self;
        jsWebViewDelegate   = webViewDelegate;
        resourceBundle      = bundle;
        bridgeHandler       = handler;
        messageHandlers     = [NSMutableDictionary dictionary];
        startupMessageQueue = [NSMutableArray array];
        responseCallbacks   = [NSMutableDictionary dictionary];
        nativeModules       = [NSMutableDictionary dictionary];
    }
    return self;
}

-(void)send:(NSString *)eventName data:(id)data responseCallback:(JSBResponseCallback)responseCallback {
    NSMutableDictionary* message = [NSMutableDictionary dictionary];
    
    message[@"status"] = @"true";
    if(data) message[@"data"] = data;
    if(eventName) message[@"eventName"] = eventName;
    
    if (responseCallback) {
        NSString* callbackId = [NSString stringWithFormat:@"objc_cb_%ld", ++uniqueId];
        responseCallbacks[callbackId] = [responseCallback copy];
        message[@"callbackId"] = callbackId;
    }
    
    [self queueMessage:message];
}

-(void)registerEvent:(NSString *)eventName handler:(JSBHandler)handler {
    messageHandlers[eventName] = [handler copy];
}
-(void)deRegisterEvent:(NSString *)eventName handler:(JSBHandler)handler {
    [messageHandlers removeObjectForKey:eventName];
}

@end

File: Android/JSBridge/assets/JSBridge.js
//
//  JSBridge.js
//  JSBridge
//
//  Created by Siva RamaKrishna Ravuri
//  Copyright (c) 2014 www.siva4u.com. All rights reserved.
//
// The MIT License (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
//

;(function(w,doc) {
    if(w.JSBridge)return;

// PRIVATE VARIABLES
    //!!! WARNING - Should be in SYNC with Native Code defines - Begin
    var JSBRIDGE_URL_SCHEME  = 'jsbridgeurlscheme';
    var JSBRIDGE_URL_MESSAGE = '__JSB_URL_MESSAGE__';
    var JSBRIDGE_URL_EVENT   = '__JSB_URL_EVENT__';
    var JSBRIDGE_URL_API     = '__JSB_URL_API__';
  
    //!!! WARNING - Should be in SYNC with Native Code defines - End
  
    var ua                  = navigator.userAgent;
    var isIOSDevice         = /iP(hone|od|ad)/g.test(ua);
    var isAndroidDevice     = /Android/g.test(ua);
    var sendMessageQueue    = [];
    var receiveMessageQueue = [];
    var messageHandlers     = {};
    var responseCallbacks   = {};
    var apiData             = null;
    var uniqueId            = 1;
    var messagingIframe;

// PRIVATE METHODS

    function JSBridgeLog() {
        if (typeof console != 'undefined') {
            console.log("JSBridge:JS: LOG: ",arguments);
        }
    }
    function JSBridgeLogException(e,m) {
        if (typeof console != 'undefined') {
            console.error("JSBridge:JS: EXCEPTION: ",arguments);
        }
    }

    function getIFrameSrc(param) {
        return JSBRIDGE_URL_SCHEME + '://' + JSBRIDGE_URL_MESSAGE + '/'+ param;
    }

    function callObjCAPI(name,data) {
        // Should not called triggerNativeCall as iFrame needs to be deleted in order to get the retvalue.
        var iframe = document.createElement("IFRAME");
        apiData = {api:name};
        if(data) apiData["data"] = data;
        iframe.setAttribute("src", getIFrameSrc(JSBRIDGE_URL_API));
        document.documentElement.appendChild(iframe);
        iframe.parentNode.removeChild(iframe);
        iframe = null;

        var ret = JSBridge.nativeReturnValue;
        JSBridge.nativeReturnValue = undefined;
        if(ret) return decodeURIComponent(ret);
    };

    function triggerNativeCall() {
        if(isIOSDevice) {
            messagingIframe.src = getIFrameSrc(JSBRIDGE_URL_EVENT);
        } else {
            var apiName = ((isAndroidDevice)?("AndroidAPI.ProcessJSEventQueue"):("WebAppAPI.ProcessJSEventQueue"));
            try {
                var api = eval(apiName);
                if(api) api(_fetchJSEventQueue());
            } catch(e) {}
        }
    }

    function doSend(message, responseCallback) {
        if (responseCallback) {
            var callbackId = 'cb_' + (uniqueId++) + '_' + new Date().getTime();
            responseCallbacks[callbackId] = responseCallback;
            message['callbackId'] = callbackId;
        }
        sendMessageQueue.push(message);
        triggerNativeCall();
    }

    function dispatchMessageFromNative(messageJSON) {
        setTimeout(function _timeoutDispatchMessageFromObjC() {
            var message = JSON.parse(messageJSON);
            var messageHandler;
            var responseCallback;

            if (message.responseId) {
                responseCallback = responseCallbacks[message.responseId];
                if(!responseCallback){return;}
                responseCallback(message.responseData);
                delete responseCallbacks[message.responseId];
            } else {
                if (message.callbackId) {
                    var callbackResponseId = message.callbackId;
                    responseCallback = function(responseData) {
                        doSend({responseId:callbackResponseId, responseData:responseData});
                    }
                }
                
                try {
                    var handler = ((message.eventName)?(messageHandlers[message.eventName]):(JSBridge.bridgeHandler));
                    if(handler) {
                    	handler(message.data, responseCallback);
                    }
                } catch(e) {
                    JSBridgeLogException(e,"dispatchMessageFromNative");
                }
            }
        });
    }

// PUBLIC METHODS
    function init(bridgeHandler) {
        if(JSBridge.bridgeHandler){JSBridgeLogException(e,"init");}
        JSBridge.bridgeHandler  = bridgeHandler;
        var receivedMessages    = receiveMessageQueue;
        receiveMessageQueue     = null;
        for(var i=0; i<receivedMessages.length; i++) {
            dispatchMessageFromNative(receivedMessages[i]);
        }
    }

    function send(eventName, data, responseCallback) {
        var dataToSend = {};
        if(eventName) dataToSend["eventName"] = eventName;
        if(data) dataToSend["data"] = data;
        doSend(dataToSend, responseCallback);
    }

    function registerEvent(eventName, handler) {
        messageHandlers[eventName] = handler;
    }

    function deRegisterEvent(eventName, handler) {
        if(messageHandlers[eventName]) {
            delete messageHandlers[eventName];
        }
    }

    function callAPI(name, data, responseCallback) {
        try {
            if(data) {
                if(responseCallback) {
                    var cbID = "cbID" + (+new Date);
                    responseCallbacks[cbID] = responseCallback;
                    data["callbackID"] = cbID;
                }
                try{data = JSON.stringify(data);}catch(e){}
            }

            if(isIOSDevice) {
                if(data) name += ":";
                return callObjCAPI(name,data);
            } else {
                var api = eval((isAndroidDevice)?("AndroidAPI.ProcessJSAPIRequest"):("WebAppAPI.ProcessJSAPIRequest"));
                if(api) {
                	if(data) return api(name,data);
                	return api(name,null);
                } else {
	                JSBridgeLogException("Unsupported API:",name);
                }
			}
        } catch(e) {
            JSBridgeLogException(e,"Invalid API:"+name);
        }
    }
  
    function _fetchJSEventQueue() {
        try {
            var messageQueueString = JSON.stringify(sendMessageQueue);
            sendMessageQueue = [];
            return messageQueueString;
        } catch(e) {
            JSBridgeLogException(e,"_fetchJSEventQueue");
        }
        return [];
    }

    function _handleMessageFromNative(messageJSON) {
        if(receiveMessageQueue) {
            receiveMessageQueue.push(messageJSON);
        } else {
            dispatchMessageFromNative(messageJSON);
        }
    }

    function _getAPIData() { return JSON.stringify(apiData); }
  
    function _invokeJSCallback(cbID,removeAfterExecute,config) {
        if(cbID) {
            var cb = responseCallbacks[cbID];
            if(cb) {
                if(removeAfterExecute) delete(responseCallbacks[cbID]);
                var data = config;
                if(isAndroidDevice) {
                    try {data = JSON.parse(config);}catch(e){}
                }
                if(data.callbackID) delete(data.callbackID);
                cb.call(null, data);
            }
        }
    };

	w.JSBridge = {
		init    : init,
		send    : send,
        callAPI : callAPI,
  
		registerEvent   : registerEvent,
        deRegisterEvent : deRegisterEvent,
  
		_fetchJSEventQueue      : _fetchJSEventQueue,
		_handleMessageFromNative: _handleMessageFromNative,
        _getAPIData             : _getAPIData,
        _invokeJSCallback       : _invokeJSCallback,
    }

    messagingIframe = doc.createElement('iframe');
    messagingIframe.style.display = 'none';
    triggerNativeCall();
    doc.documentElement.appendChild(messagingIframe);

    var readyEvent = doc.createEvent('Events');
    readyEvent.initEvent('JSBridgeReady');
    readyEvent.bridge = JSBridge;
    doc.dispatchEvent(readyEvent);

})(window,document);
